core/include/vecmem/containers/device_vector.hpp

```diff
@@ -35,10 +36,15 @@ namespace vecmem {
       /// Type of the array elements
       typedef TYPE              value_type;
       /// Size type for the array
-      typedef std::size_t       size_type;
+      typedef unsigned int      size_type;
       /// Pointer difference type
       typedef std::ptrdiff_t    difference_type;
 
+      /// Pointer type to the size of the array
+      typedef typename std::conditional< std::is_const< TYPE >::value,
+                                         const size_type*,
+                                         size_type* >::type size_pointer;
+
```

Comment:
I don't really understand what the constness of `size_pointer` is relevant for. Could you elaborate?

Reply:
This is a tricky little thing... With the device code the const-ness of something can not be enforced through the const-ness of the "high level" objects. Since we always pass `vecmem::data::vector_view` types by value to kernels (that's the only way of doing it really...), you can't express that a vector should be const, through an "external" `const` declaration.

What you can do instead is to use types like `vecmem::data::vector_view<const int>` to show that this is a view of a constant `int` array. (This is why `vecmem::const_device_vector` "is a thing".)

Now... Here I was thinking that the const-ness of the elements should also extend to the const-ness of the size of the vector. I.e. if you pass a `vecmem::data::vector_view<const T>` object to your kernel, that should mean that not only do you not want to allow the kernel to modify the values of the elements in the vector, but also that the kernel should not be allowed to resize the vector.

And that's what I tried to express with this piece of code. :wink:

Reply:
I see your point. Just from looking at it, I would expect `vecmem::data::vector_view<const T>` to let me change the size though. But ok.